### PR TITLE
Move ort flatbuffers helper functions and value info r/w functions into separated lib

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Move ort flatbuffers helper functions and value info r/w functions into separated lib

**Motivation and Context**
- Remove the dependency of graph/model if we want to get input/output info from ort model
- Remove ort flatbuffers header from graph/model/ssession_state headers
- Remove many static linked libs for onnxruntime_perf_test
- Enable ort model support for full build onnx_test_runner/onnxruntime_perf_test

### CODE REVIEW

File: onnxruntime/core/flatbuffers/flatbuffers_utils.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <core/common/status.h>
+#include <core/graph/basic_types.h>
+#include <core/session/onnxruntime_c_api.h>
+
```
Comment (skottmckay) [inline onnxruntime/core/flatbuffers/flatbuffers_utils.h:9]: the C API is a really high level thing. why does a low level thing like this library have a dependency on it?

File: cmake/onnxruntime_flatbuffers.cmake [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+add_library(onnxruntime_flatbuffers ${onnxruntime_flatbuffers_srcs})
+install(DIRECTORY ${PROJECT_SOURCE_DIR}/../include/onnxruntime/core/flatbuffers  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/onnxruntime/core)
+onnxruntime_add_include_to_target(onnxruntime_flatbuffers onnx flatbuffers)
+if(onnxruntime_ENABLE_INSTRUMENT)
+  target_compile_definitions(onnxruntime_flatbuffers PUBLIC ONNXRUNTIME_ENABLE_INSTRUMENT)
+endif()
+target_include_directories(onnxruntime_flatbuffers PRIVATE ${ONNXRUNTIME_ROOT})
+add_dependencies(onnxruntime_flatbuffers ${onnxruntime_EXTERNAL_DEPENDENCIES})
+set_target_properties(onnxruntime_flatbuffers PROPERTIES FOLDER "ONNXRuntime")
```
Comment (skottmckay) [inline cmake/onnxruntime_flatbuffers.cmake:19]: this has a dependency on the graph library doesn't it given flatbuffers_utils.h includes a header from there?

File: onnxruntime/core/flatbuffers/flatbuffers_utils.h (hunk 1 above)
Comment (gwang-msft) [reply onnxruntime/core/flatbuffers/flatbuffers_utils.h:9]: Oh, this include is only for the "ORT_MUST_USE_RESULT" which is defined in the c_api

---
In reply to: [494752637](https://github.com/microsoft/onnxruntime/pull/5276#discussion_r494752637) [](ancestors = 494752637)

File: cmake/onnxruntime_flatbuffers.cmake (hunk 2 above)
Comment (skottmckay) [reply cmake/onnxruntime_flatbuffers.cmake:19]: may not need a link dependency but would be good to be more explicit about the dependency on the include path

---
In reply to: [494752823](https://github.com/microsoft/onnxruntime/pull/5276#discussion_r494752823) [](ancestors = 494752823)

Comment (gwang-msft) [reply cmake/onnxruntime_flatbuffers.cmake:19]: Including  core/graph/basic_types.h is only for the declaration of ONNX_NAMESPACE::ValueInfoProto, removed and add some forward decls



---
In reply to: [494759244](https://github.com/microsoft/onnxruntime/pull/5276#discussion_r494759244) [](ancestors = 494759244,494752823)

File: orttraining/orttraining/core/graph/optimizer_graph_builder.cc [hunk 3]
```diff
@@ -256,7 +256,7 @@ Status OptimizerGraphBuilder::AddGradientNorm(
                   "Unsupport gradient type: it has to be either float, MLFloat16 or BFloat16.");
   }
 
-  for (const auto argdef : grad_argdefs) {
+  for (const auto& argdef : grad_argdefs) {
     ONNX_NAMESPACE::TensorProto_DataType elem_type =
```
Comment (gwang-msft) [inline orttraining/orttraining/core/graph/optimizer_graph_builder.cc:260]: This is a minor fix for a warning on latest Apple clang

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]
:shipit:

**Comment by xkszltl** [comment]
@gwang-msft FYI this PR breaks master build.
https://github.com/microsoft/onnxruntime/commit/3a3f26f38e049eb0e8b9fbb97281a6e5dd82074f#commitcomment-42736525

**Comment by gwang-msft** [comment]
@xkszltl, yes, this will cause problems, not sure why CI did not catch this.
What's you build command? I will fix this

**Comment by xkszltl** [comment]
I uses ninja, maybe the generator is more sensitive than makefile.

**Comment by xkszltl** [comment]
This is the build arg: https://github.com/xkszltl/Roaster/blob/14e68367389ad0b8fd809a48b4b70a76f183f89e/pkgs/ort.sh#L75-L135
Most of it should be irrelevant to this issue.

**Comment by xkszltl** [comment]
That was another one, so it'll be great if it can be captured by CI:
https://github.com/microsoft/onnxruntime/issues/5024

**Comment by gwang-msft** [comment]
This https://github.com/microsoft/onnxruntime/pull/5306 should fix the issue

**Pull request merged**